lib/elixir/lib/regex.ex

```diff
@@ -677,7 +677,7 @@ defmodule Regex do
     <<_::binary-size(^offset), part::binary-size(^keep), match::binary-size(^length), _::binary>> =
       string
 
-    if keep == 0 and trim do
+    if keep == 0 and (trim or (offset != 0 and length == 0)) do
```

Comment:
You may need this to be a cond:

```
    cond do
      keep == 0 and offset != 0 and length == 0 ->
        do_split([h | t], string, new_offset, counter - 1, trim, true)

      keep == 0 and trim ->
        [match | do_split([h | t], string, new_offset, counter - 1, trim, true)]

      true ->
        [part, match | do_split([h | t], string, new_offset, counter - 1, trim, true)]
    end
```

I haven't tested it but it may fix my additional reports below.

Reply:
@josevalim [be37140](https://github.com/elixir-lang/elixir/pull/14468/commits/be3714038dfd6c2320cbd7a564e8943952e11a87)